applications/GeoMechanicsApplication/custom_workflows/dgeosettlement.cpp

```diff
@@ -31,44 +31,60 @@ KratosGeoSettlement::KratosGeoSettlement()
     }
 }
 
-int KratosGeoSettlement::RunStage(const std::string&                      rWorkingDirectory,
-                                  const std::string&                      rProjectParametersFileName,
+int KratosGeoSettlement::RunStage(const std::filesystem::path&            rWorkingDirectory,
+                                  const std::filesystem::path&            rProjectParametersFile,
                                   const std::function<void(const char*)>& ,
                                   const std::function<void(double)>&      ,
                                   const std::function<void(const char*)>& ,
                                   const std::function<bool()>&            )
 {
     KRATOS_INFO("KratosGeoSettlement") << "About to run a stage..." << std::endl;
 
-    const auto project_parameters_file_path = rWorkingDirectory + "/" + rProjectParametersFileName;
-    const auto project_parameters = InputUtilities::ProjectParametersFrom(project_parameters_file_path);
+    const auto project_parameters_file_path = rWorkingDirectory / rProjectParametersFile;
+    const auto project_parameters = InputUtilities::ProjectParametersFrom(project_parameters_file_path.generic_string());
     KRATOS_INFO("KratosGeoSettlement") << "Parsed project parameters file " << project_parameters_file_path << std::endl;
 
-    const auto model_part_name = project_parameters["solver_settings"]["model_part_name"].GetString();
-    ModelPart& model_part = mModel.CreateModelPart(model_part_name);
-    model_part.SetBufferSize(2);
-    KRATOS_INFO("KratosGeoSettlement") << "Created a model part" << std::endl;
+    if (const auto model_part_name = project_parameters["solver_settings"]["model_part_name"].GetString();
+        !mModel.HasModelPart(model_part_name)) {
+        auto& model_part = AddNewModelPart(model_part_name);
+        const auto mesh_file_name = project_parameters["solver_settings"]["model_import_settings"]["input_filename"].GetString();
+        ReadModelFromFile(rWorkingDirectory / mesh_file_name, model_part);
+    }
+
+    if (project_parameters["solver_settings"].Has("material_import_settings")) {
+        const auto material_file_name = project_parameters["solver_settings"]["material_import_settings"]["materials_filename"].GetString();
+        const auto material_file_path = rWorkingDirectory / material_file_name;
+        InputUtilities::AddMaterialsFrom(material_file_path.generic_string(), mModel);
+        KRATOS_INFO("KratosGeoSettlement") << "Read the materials from " << material_file_path << std::endl;
+    }
+
+    return 0;
+}
+
+ModelPart& KratosGeoSettlement::AddNewModelPart(const std::string& rModelPartName)
+{
+    auto& result = mModel.CreateModelPart(rModelPartName);
```

Comment:
Are we sure that none of these steps need to be 'redone' (i.e. are there any steps here that change something that's not persistent in between steps?)

Reply:
Yes, I think we can be rather confident that we need to carry out these actions only once per model part. If we come across a case where this doesn't apply, we probably also need to modify the corresponding Python workflow.